src/censored.jl

```diff
@@ -0,0 +1,519 @@
+    return promote_type(partype(d.uncensored), T)
+end
+
+Base.eltype(::Type{<:Censored{D,S,T}}) where {D,S,T} = promote_type(T, eltype(D))
+
+#### Range and Support
+
+islowerbounded(d::LeftCensored) = islowerbounded(d.uncensored)
+islowerbounded(d::Censored) = islowerbounded(d.uncensored) || cdf(d.uncensored, d.lower) > 0
+
+isupperbounded(d::RightCensored) = isupperbounded(d.uncensored)
+isupperbounded(d::Censored) = isupperbounded(d.uncensored) || _ccdf_inclusive(d.uncensored, d.upper) > 0
+
+minimum(d::LeftCensored) = min(minimum(d.uncensored), d.upper)
+minimum(d::Censored) = max(minimum(d.uncensored), d.lower)
+
+maximum(d::RightCensored) = max(maximum(d.uncensored), d.lower)
+maximum(d::Censored) = min(maximum(d.uncensored), d.upper)
+
+function insupport(d::Censored, x::Real)
+    d0 = d.uncensored
+    lower = d.lower
+    upper = d.upper
+    return (
+        (_in_open_interval(x, lower, upper) && insupport(d0, x)) ||
+        (x == lower && cdf(d0, lower) > 0) ||
+        (x == upper && _ccdf_inclusive(d0, upper) > 0)
+    )
+end
+
+#### Show
+
+function show(io::IO, ::MIME"text/plain", d::Censored)
+    print(io, "Censored(")
+    d0 = d.uncensored
+    uml, namevals = _use_multline_show(d0)
+    uml ? show_multline(io, d0, namevals; newline=false) : show_oneline(io, d0, namevals)
+    if d.lower === nothing
+        print(io, "; upper=$(d.upper))")
+    elseif d.upper === nothing
+        print(io, "; lower=$(d.lower))")
+    else
+        print(io, "; lower=$(d.lower), upper=$(d.upper))")
+    end
+end
+
+_use_multline_show(d::Censored) = _use_multline_show(d.uncensored)
+
+
+#### Statistics
+
+quantile(d::Censored, p::Real) = _clamp(quantile(d.uncensored, p), d.lower, d.upper)
+
+median(d::Censored) = _clamp(median(d.uncensored), d.lower, d.upper)
+
+# the expectations use the following relation:
+# 𝔼_{x ~ d}[h(x)] = P_{x ~ d₀}(x < l) h(l) + P_{x ~ d₀}(x > u) h(u)
+#                 + P_{x ~ d₀}(l ≤ x ≤ u) 𝔼_{x ~ τ}[h(x)],
+# where d₀ is the uncensored distribution, d is d₀ censored to [l, u],
+# and τ is d₀ truncated to [l, u]
+
+function mean(d::LeftCensored)
+    upper = d.upper
+    log_prob_upper = logccdf(d.uncensored, upper)
+    prob_upper = exp(log_prob_upper)
+    μ = prob_upper * (iszero(prob_upper) ? oneunit(upper) : upper)
+    prob_trunc = exp(log1mexp(log_prob_upper))
+    # truncation contains ~ no probability
+    prob_trunc < eps(one(prob_trunc)) && return μ
+    return μ + oftype(μ, prob_trunc * mean(_to_truncated(d)))
+end
+function mean(d::RightCensored)
+    lower = d.lower
+    log_prob_lower = _logcdf_noninclusive(d.uncensored, lower)
+    prob_lower = exp(log_prob_lower)
+    μ = prob_lower * (iszero(prob_lower) ? oneunit(lower) : lower)
+    prob_trunc = exp(log1mexp(log_prob_lower))
+    # truncation contains ~ no probability
+    prob_trunc < eps(one(prob_trunc)) && return μ
+    return μ + oftype(μ, prob_trunc * mean(_to_truncated(d)))
```

Comment:
Same here, maybe
```suggestion
    return μ + oftype(μ, prob_trunc * mean(_to_truncated(d)))
```
?